Added support to create variants on object creation

turbo-ele opened the pull request:
Allows the import of variants on `Element Creation` and `Element Location Update`, when using the `findParent` location strategy.

The type of the element will then be set to variant if all of the following conditions apply:
- Newly added `As Variant` checkbox is checked
- Element needs to be of the same class as the parent
- The class needs to allow the creation of variants
- Element is not allowed to have children

Additionally two more variant related checks were added:
- at the findParent strategy only parents can be set that are not of type `variant`
- generally, the parent of a variant cannot be changed anymore

turbo-ele committed:
Added support to create variants on object creation

turbo-ele committed:
Added support to create variants on object creation

blankse commented on src/Resolver/Location/FindParentStrategy.php:
```diff
@@ -57,6 +57,8 @@ class FindParentStrategy implements LocationStrategyInterface
      */
     protected $attributeLanguage;
 
+    protected bool $saveAsVariant;
```
```suggestion
    protected bool $saveAsVariant = false;
```

blankse commented on src/Resolver/Location/FindParentStrategy.php:
```diff
@@ -75,6 +77,8 @@ public function setSettings(array $settings): void
             throw new InvalidConfigurationException('Empty find strategy.');
         }
 
+        $this->saveAsVariant = !empty($settings['asVariant']) && $settings['asVariant'] === 'on';
```
```suggestion
        $this->saveAsVariant = isset($settings['asVariant']) && $settings['asVariant'] === 'on';
```

blankse reviewed (commented):
Small improvements

turbo-ele committed:
Apply suggestions from code review

Co-authored-by: Sebastian Blank <sebastian.bl@gmx.de>

turbo-ele committed:
Added error handling when changing existing variants

As the parent can be changed for variants this is now catched, as a variant cannot change its
parent anymore.

turbo-ele committed:
Added error handling when changing existing variants

As the parent can be changed for variants this is now catched, as a variant cannot change its
parent anymore.

fashxp commented:
can you please just add the english translation. all others are maintained via poeditor. thx very much.

turbo-ele committed:
Removed unecessary translations

turbo-ele committed:
Addressed some qodana warnings

fashxp commented on src/Resolver/Resolver.php:
```diff
@@ -149,7 +154,18 @@ public function loadOrCreateAndPrepareElement(array $inputData, bool $createNew
             $this->getLocationUpdateStrategy()->updateParent($element, $inputData);
+
+            // The parent of a variant cannot be changed anymore.
+            if (
```
I'm just wondering, if we could pack also that check into the FindParentStrategy? 
Then only the FindParentStrategy needs to know about the variant handling...

fashxp replied:
@turbo-ele :ping_pong:  any thoughts on this?

turbo-ele replied:
@fashxp sorry, somehow I missed the first message.
I would say moving it into the strategy is not an option, as a variant - correct me if I'm wrong - should never be moved from it's parent. So if we'd move it into the FindParentStrategy, another location update strategy, such as the StaticPathStrategy, would otherwise be able to move the variant into a folder.
Alternatively, this check has to be implemented in every strategy individually, which i wanted to avoid.

fashxp replied:
you are completely right!

fashxp commented on src/Resolver/Location/FindParentStrategy.php:
```diff
@@ -129,7 +136,28 @@ public function updateParent(ElementInterface $element, array $inputData): Eleme
 
         if ($newParent) {
-            return $element->setParent($newParent);
+            if ($newParent->getType() === AbstractObject::OBJECT_TYPE_VARIANT) {
```
in UI we allow variants of variants, so I think we should not restrict that here either.

turbo-ele replied:
Right, I did overlook this. So I now added the condition, that a variant can be set as a parent of a variant, if it's of the same class as the element itself.

fashxp replied:
sorry for being picky, but that condition actually applies for all parents .. a variant always has to be of the same type as its parent, regardless if it is an object or a variant.

turbo-ele replied:
It's fine if you're being picky, I actually appreciate that 😃 

But I'm not entirely sure, if I get you right here. A variant needs to be of the same class as its parent - this is what I actually want to verify here. Let me try elaborate what my intention was - and you can - of course - correct me if I'm wrong.

I didn't want it to be possible to set a parent to any object, if that parent is a variant, unless the element itself is a variant and it's of the same class as the parent. Otherwise you could assign non variant elements to a parent that is a variant, or assign variants to a variant of another class.

https://github.com/pimcore/data-importer/blob/830fef0deb5bb29098e6ebee2b7394a034ab671a/src/Resolver/Location/FindParentStrategy.php#L139-L149

First, I'm checking if the parent is a variant. If this is the case, the element needs to be a variant of the same class as well. Otherwise, I could either assign a variant of another class - or any other class types to a variant - which is also not possible in the admin-ui.

Am I missing something? Or is there already a built in functionality checking this?
If you think the import should allow that, I'm also totally fine with that and I will remove the entire condition.


---

Another thing I just realized is that the following condition also is not working for variants having variants:
https://github.com/pimcore/data-importer/blob/830fef0deb5bb29098e6ebee2b7394a034ab671a/src/Resolver/Location/FindParentStrategy.php#L157
I'm biased on this one. We have the following possibilities:
1. you can only change an object to a variant if it hasn't set any children yet, this also has the advantage that it won't be possible to create a tree structure where no parent of type object exists.
1. we check for all children, if they are variants of the object - that sounds a little performance hungry and imo should be avoided
1. or we could ignore that and simply make it possible to change an object to a variant regardless if it has any children or not

I actually prefer the current solution (1)

fashxp replied:
I get your point. But there is still an inconsistency. 
1) Trying to assign a variant of different class below a variant leads to an exception, because of https://github.com/pimcore/data-importer/pull/375/files#diff-8f9965171cbd95a20bc79746f27b5472c3243e89c9b0e865fdc2618168c25a39R145-R149
2) Trying to assign a variant of different class below an object is just silently omitted by not creating a variant but an normal object, because of https://github.com/pimcore/data-importer/pull/375/files#diff-8f9965171cbd95a20bc79746f27b5472c3243e89c9b0e865fdc2618168c25a39R158-R164

Or do I miss something? 

I would say, we should throw exceptions in all cases when creating/updating a variant is not possible but configured.

fashxp replied:
why is `!$element->hasChildren() ` causing problems with variants having variants? 

but I'm with you, we should allow converting to variants only for objects without any children (what this condition actually does, right?).

turbo-ele replied:
> why is `!$element->hasChildren() ` causing problems with variants having variants?
> 

I was thinking about the case where you have
```
object (1)
|- object (2)
|-- object (3)
```
with the `!$element->hasChildren() ` check you cannot change `object (2)` to a variant, although it might be legit.
```
object (1)
|- variant (2) <- this doesn't work, as it has a child
|-- object (3)
```

> but I'm with you, we should allow converting to variants only for objects without any children (what this condition actually does, right?).

Exactly. So we'll leave it like that.

turbo-ele replied:
>I would say, we should throw exceptions in all cases when creating/updating a variant is not possible but configured.

Ok, I will create an exception for any case

fashxp reviewed (changes requested):
I would do one minor adaption, besides that it looks good to me :+1:

turbo-ele committed:
Allowing variants to be parents of variants of the same class

turbo-ele committed:
Apply php-cs-fixer changes

turbo-ele committed:
fixup! Allowing variants to be parents of variants of the same class

turbo-ele committed:
Merge branch 'feature/support-variants' of github.com:turbo-ele/data-importer into feature/support-variants

turbo-ele committed:
Added check if parent has changed

turbo-ele commented on src/Resolver/Location/FindParentStrategy.php:
```diff
@@ -128,8 +135,40 @@ public function updateParent(ElementInterface $element, array $inputData): Eleme
             $newParent = DataObject::getByPath($this->fallbackPath);
         }
 
+        // If parent has not changed - no need to update.
+        if ($element->getParentId() && $element->getParentId() === $newParent?->getId()) {
+            return $element;
+        }
+
```
@fashxp - and i just added this condition. As all the checks can be omitted if the parent hasn't changed at all

fashxp replied:
But then, an objects type could not be changed (to variant/object) anymore during an update. This might cause confusion, as the `as variant` checkbox is also there for updating parent. 

I would still allow that with the restrictions we are discussing below.

turbo-ele replied:
Ah, good catch. You're right, I just disabled the type change on updates. 😕 

The problem i tried to solve is on the one hand, that 
- we don't have to check the status quo, as well as that
- the condition below can otherwise throw an exception on updates, if the element that we're trying to update might already be a child of a variant (e.g. it could have been moved via code below a variant - this is currently possible). If we'd only perform this check when the parent changes, then it would not be a problem.
https://github.com/pimcore/data-importer/blob/e0a48f47714d74251e47598e5bc987270e9bc794/src/Resolver/Location/FindParentStrategy.php#L144-L150
Of course you can argue, that this should actually never be the case as it's not intended to have children on variants other than variants. Although Pimcore doesn't keep you from doing this 


Here's an example

```
product
|- variant
|-- price-object   <- this cannot be updated anymore, as it's parent is a variant 
```

fashxp replied:
I see. I would say lets accept this potential problem for now and see if it really becomes one.

turbo-ele committed:
Added individual exceptions when variant type cannot be set

turbo-ele committed:
Apply php-cs-fixer changes

turbo-ele committed:
fixup! Added individual exceptions when variant type cannot be set

turbo-ele committed:
Merge branch 'feature/support-variants' of github.com:turbo-ele/data-importer into feature/support-variants

fashxp commented on src/Resolver/Location/FindParentStrategy.php:
```diff
@@ -138,4 +159,50 @@ public function updateParent(ElementInterface $element, array $inputData): Eleme
+    {
+        // Check if element should be saved as a variant if not already.
+        if (
+            $this->saveAsVariant && $element->getType() !== AbstractObject::OBJECT_TYPE_VARIANT
```
Now, another case came into my mind:
If I have a `variant`, I cannot change the type to `object` anymore. The other way round it is possible. 
I would say, either we rename the option 'force variant', or we allow changing both ways (with the implemented restrictions).
WDYT?

fashxp replied:
@turbo-ele :ping_pong:

turbo-ele replied:
you are right, i would say it should be changeable in both ways. i'm going to change it

fashxp replied:
@turbo-ele can you have a look at phpstan.

turbo-ele committed:
Allowing variants to be changed to objects

turbo-ele committed:
fixup! Allowing variants to be changed to objects

fashxp reviewed (approved):
finally it looks good. thank you so much for your support! this is a great addition.
